monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/Pad2Observable.scala

```diff
@@ -0,0 +1,191 @@
+package monix.reactive.internal.builders
+
+/**
+  * Created by shengc on 6/21/16.
+  */
+/*
+ * Copyright (c) 2014-2016 by its authors. Some rights reserved.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import monix.execution.cancelables.CompositeCancelable
+import monix.execution.{Cancelable, Ack}
+import monix.execution.Ack.{Stop, Continue}
+import monix.reactive.Observable
+import monix.reactive.observers.Subscriber
+import scala.concurrent.{Promise, Future}
+import scala.util.Success
+import scala.util.control.NonFatal
+
+private[reactive] final
+class Pad2Observable[A1,A2,+R]
+(obsA1: Observable[A1], obsA2: Observable[A2], paddingA1: ⇒ A1, paddingA2: ⇒ A2)
+(f: (A1,A2) ⇒ R)
+  extends Observable[R] { self ⇒
+
+  def unsafeSubscribeFn(out: Subscriber[R]): Cancelable = {
+    import out.scheduler
+
+    // MUST BE synchronized by `self`
+    var isDone = false
+    // MUST BE synchronized by `self`
+    var lastAck = Continue : Future[Ack]
+    // MUST BE synchronized by `self`
+    var elemA1 = null.asInstanceOf[A1]
+    // MUST BE synchronized by `self`
+    var hasElemA1 = false
+    // MUST BE synchronized by `self`
+    var elemA2 = null.asInstanceOf[A2]
+    // MUST BE synchronized by `self`
+    var hasElemA2 = false
+    // MUST BE synchronized by `self`
+    var continueP = Promise[Ack]()
+    // MUST BE synchronized by `self`
+    var completedCount = 0
+
+    // MUST BE synchronized by `self`
+    def rawOnNext(a1: A1, a2: A2): Future[Ack] =
+      if (isDone) Stop else {
+        var streamError = true
+        try {
+          val c = f(a1,a2)
+          streamError = false
+          out.onNext(c)
+        } catch {
+          case NonFatal(ex) if streamError =>
+            isDone = true
+            out.onError(ex)
+            Stop
+        } finally {
+          hasElemA1 = false
+          hasElemA2 = false
+        }
+      }
+
+    // MUST BE synchronized by `self`
+    def signalOnNext(a1: A1, a2: A2): Future[Ack] = {
+      lastAck = lastAck match {
+        case Continue => rawOnNext(a1,a2)
+        case Stop => Stop
+        case async =>
+          async.flatMap {
```

Comment:
I'm not sure if this back-pressuring here is necessary. We have to think about it, but here's my rationale:

1. on the first item only `A1` is allowed to emit, which means `continueA2` is on `Continue` from the start
2. we emit and then we say `continueA1.tryCompleteWith(lastAck)` 
3. and so `A2` will emit only when `continueA1` is complete and this implies that our `lastAck` here is already completed

Also when you store an `ack` and then do an `ack.flatMap`, many times it is better to do `ack.syncTryFlatten.syncFlatMap`, because otherwise a single asynchronous `ack` can introduce asynchronous boundaries on all the remaining events and that's very inefficient.